staging/src/k8s.io/apiserver/pkg/util/shufflesharding/shufflesharding.go

```diff
@@ -0,0 +1,115 @@
+// in order to reduce this unevenness to 32:33, we set MaxHashBits to 60 here.
+const MaxHashBits = 60
+
+// RequiredEntropyBits will figure out the required entropy bits from the given
+// deckSize and handSize by calculating bits of (deckSize^handSize).
+// RequiredEntropyBits must be less or equal than MaxHashBits
+// if we want to create and use Dealer.
+func RequiredEntropyBits(deckSize, handSize int) int {
+	return big.NewInt(1).Exp(big.NewInt(int64(deckSize)), big.NewInt(int64(handSize)), nil).BitLen()
+}
+
+// Dealer contains some necessary parameters and provides some methods for shuffle sharding.
+// Dealer is thread-safe.
+type Dealer struct {
+	deckSize int
+	handSize int
+
+	// rawCardsPool is a pool for caching and reusing rawCards slices to improve performance
+	rawCardsPool *sync.Pool
+}
+
+// NewDealer will create a Dealer with the given deckSize and handSize, will return error when
+// deckSize or handSize is invalid as below.
+// 1. deckSize or handSize is not positive
+// 2. handSize is greater than deckSize
+// 3. required entropy bits of deckSize and handSize is greater than MaxHashBits
+func NewDealer(deckSize, handSize int) (*Dealer, error) {
+	if deckSize <= 0 || handSize <= 0 {
+		return nil, fmt.Errorf("deckSize %d or handSize %d is not positive", deckSize, handSize)
+	}
+	if handSize > deckSize {
+		return nil, fmt.Errorf("handSize %d is greater than deckSize %d", handSize, deckSize)
+	}
+	if RequiredEntropyBits(deckSize, handSize) > MaxHashBits {
+		return nil, fmt.Errorf("required entropy bits of deckSize %d and handSize %d is greater than %d", deckSize, handSize, MaxHashBits)
+	}
+
+	return &Dealer{
+		deckSize: deckSize,
+		handSize: handSize,
+		rawCardsPool: &sync.Pool{
+			New: func() interface{} {
+				return make([]int, handSize)
+			},
+		},
+	}, nil
+}
+
+// Deal shuffles a card deck of the given size and deals a hand of the given size,
+// using the given hashValue as a source of entropy.
+// This function makes handSize sequential calls to pick, one for each dealt card.
+// Each card is identified by an integer in the range [0, deckSize).
+// For example, for deckSize=128 and handSize=4 this function might call pick(14); pick(73); pick(119); pick(26).
+// This function can be extended with different pick functions.
+func (d *Dealer) Deal(hashValue uint64, pick func(int)) {
+	rawCards := d.rawCardsPool.Get().([]int)
+	defer d.rawCardsPool.Put(rawCards)
+
+	for i := 0; i < d.handSize; i++ {
+		hashValueNext := hashValue / uint64(d.deckSize-i)
+		rawCards[i] = int(hashValue - uint64(d.deckSize-i)*hashValueNext)
+		hashValue = hashValueNext
+	}
+
+	for i := 0; i < d.handSize; i++ {
+		card := rawCards[i]
+		for j := i; j > 0; j-- {
+			if card >= rawCards[j-1] {
+				card++
+			}
+		}
+		pick(card)
+	}
+}
+
+// DealIntoHand shuffles a deck of the given size by the
+// given hash value and deals cards into a hand(slice).
+func (d *Dealer) DealIntoHand(hashValue uint64) []int {
```

Comment:
Remember https://github.com/kubernetes/kubernetes/blob/feature-rate-limiting/staging/src/k8s.io/apiserver/pkg/util/shufflesharding/shufflesharding.go#L108

DealIntoHand should take the slice as a parameter, not return the slice.

Reply:
+1

Reply:
Thanks for catching this, I'll update this together after we come to a conclusion about the validations and rawCards above...